src/Avalonia.Input/KeyboardDevice.cs

```diff
@@ -31,6 +31,63 @@ private set
                 RaisePropertyChanged();
             }
         }
+        
+        private void ClearFocusWithin(IInputElement element, bool clearRoot)
+        {
+            foreach (IInputElement el in element.VisualChildren)
+            {
+                if (el.IsKeyboardFocusWithin)
+                {
+                    ClearFocusWithin(el, true);
+                    break;
+                }
+            }
+            
+            if(clearRoot)
+            {
+                if (element is InputElement ie)
+                {
+                    ie.IsKeyboardFocusWithin = false;
+                }
+            }
+        }
+
+        private void SetIsFocusWithin(IInputElement oldElement, IInputElement newElement)
```

Comment:
I believe :focus-within should work with any kind of input, not only Keyboard:
https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within

Reply:
If I have understood correctly... KeyboardFocus can be triggered by mouse click or touch press inside a control... and Focused means the control that is currently receiving keyboard input.

Its this class that is the source of controls becoming focused.

@grokys might have some input, here. I think its already compatible with any input type?